### PR TITLE
[sonic-utilities] managementVRF cli support(l3mdev)

### PR DESCRIPTION
**Pull request by vharish02**
This commit adds CLI support for management VRF using l3dev. mVRF can
be enabled using config vrf add mgmt and deleted using config vrf del mgmt.
Show commands for management VRF are added which displays the Linux command
output, will update show command display after concluding what would be the
output for the show commands.
Added cli to configure management interface(eth0), config interface ip eth0
add can be used to configure eth0 ip and config ip eth0 remove is used to
remove eth0 ip.

New cli config/show commands:

```
config vrf add mgmt
config vrf del mgmt
config interface eth0 ip add ip/mask gatewayIP
config interface eth0 ip remove ip/mask

show mgmt-vrf
show mgmt-vrf route
show mgmt-vrf addresses
show mgmt-vrf interfaces
```

Signed-off-by: Harish Venkatraman <harish_venkatraman@dell.com>

**- What I did**
Added CLI support for management VRF
**- How I did it**
Added following new CLI commands to enable and disable management VRF using l3mdev and configure eth0 management interface ip address.

```
config commands:
config vrf add mgmt
config vrf del mgmt
config interface ip add eth0 ip/mask gatewayIP
config interface ip remove eth0 ip/mask

show commands:
show mgmt-vrf 
show mgmt-vrf routes
```

Modified show ntp command

**- How to verify it**
Configure management VRF using the above config command and use the show commands to display the output of mgmt-vrf interfaces, addresses, route and ntp.

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

-->
**SAMPLE OUTPUTS OF SHOW COMMANDS**  
**show mgmt-vrf**  
```
root@sonic:/etc/init.d# show mgmt-vrf 

ManagementVRF : Enabled

Management VRF interfaces in Linux:
348: mgmt: <NOARP,MASTER,UP,LOWER_UP> mtu 65536 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether f2:2a:d9:bc:e8:f0 brd ff:ff:ff:ff:ff:ff
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master mgmt state UP mode DEFAULT group default qlen 1000
    link/ether 4c:76:25:f4:f9:f3 brd ff:ff:ff:ff:ff:ff
350: lo-m: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue master mgmt state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether b2:4c:c6:f3:e9:92 brd ff:ff:ff:ff:ff:ff
root@sonic:/etc/init.d#
```
**show mgmt-vrf routes**  
```
root@sonic:/etc/init.d# show mgmt-vrf routes

Routes in Management VRF Routing Table:
default via 10.16.210.254 dev eth0 metric 201 
broadcast 10.16.210.0 dev eth0 proto kernel scope link src 10.16.210.75 
10.16.210.0/24 dev eth0 proto kernel scope link src 10.16.210.75 
local 10.16.210.75 dev eth0 proto kernel scope host src 10.16.210.75 
broadcast 10.16.210.255 dev eth0 proto kernel scope link src 10.16.210.75 
broadcast 127.0.0.0 dev lo-m proto kernel scope link src 127.0.0.1 
127.0.0.0/8 dev lo-m proto kernel scope link src 127.0.0.1 
local 127.0.0.1 dev lo-m proto kernel scope host src 127.0.0.1 
broadcast 127.255.255.255 dev lo-m proto kernel scope link src 127.0.0.1 
root@sonic:/etc/init.d#
```
**show management_interface address**  
```
root@sonic:/etc/init.d# show management_interface address 
Management IP address = 10.16.210.75/24
Management NetWork Default Gateway = 10.16.210.254
root@sonic:/etc/init.d#
```

### CODE REVIEW

**Comment by a-barboza** [comment]
What relationship exists between managementVRF and VRF feature ?
There is a VRF feature which has its CLI. (https://github.com/Azure/SONiC/blob/master/doc/sonic-vrf-hld.md). Are there any CLIs which are common ? 
How can i add the VRF folks to this review ?

**Comment by prsunny** [comment]
> What relationship exists between managementVRF and VRF feature ?
> There is a VRF feature which has its CLI. (https://github.com/Azure/SONiC/blob/master/doc/sonic-vrf-hld.md). Are there any CLIs which are common ?
> How can i add the VRF folks to this review ?

@a-barboza, these are two different features implemented separately. This CLI is being reviewed by the VRF work-group and checked for any commonalities.

**Comment by a-barboza** [comment]
> > What relationship exists between managementVRF and VRF feature ?
> > There is a VRF feature which has its CLI. (https://github.com/Azure/SONiC/blob/master/doc/sonic-vrf-hld.md). Are there any CLIs which are common ?
> > How can i add the VRF folks to this review ?
> 
> @a-barboza, these are two different features implemented separately. This CLI is being reviewed by the VRF work-group and checked for any commonalities.

@prsunny Looks like there are quire a few commonalities (Eg: config vrf add, ...). So it is likely the patch will be redone.

**Comment by kannankvs** [comment]
@jleveque : Resolved the conflict. Request you to kindly review and merge it.

**Comment by kannankvs** [comment]
@prsunny  and  @jleveque : Anything else required on this? Or, can it be merged?

**Comment by a-barboza** [comment]
Is there a location where the table # of 5000 use for mgmt-vrf is mentioned ?  I checked the mgmt-vrf HLD document but could not find a reference other than "1" in the example linux commands.

**Comment by kannankvs** [comment]
@a-barboza : We shall update the document once if the PR is merged. Just waiting to see if there could be any changes in the CLI before merging based on any review comments.

**Comment by kannankvs** [comment]
@a-barboza : Some of your questions from PR472 that are relevant for this PR are answered here.

1) Why restart NTP ? (These changes are for SNMP ?)
<<Kannan>> This is not part of SNMP PR. This is part of this 463 PR. Once if 463 is merged, 472 PR will contain only mvrf SNMP CLI changes.

2) clear_mgmt command: There are some new commands common to all mgmt-vrf ?
MGMT_INTERFACE, MGMT_VRF_CONFIG tables ?
<<Kannan>> This command is an easy way to clear both the tables. 

3) There is a new "--gw" option? Why is this only for eth0 ? Can Front Panel (In-Band) port be part of the mgmt-vrf ?
<<Kannan>> As of now, eth0 alone will be part of mvrf; not front panel ports. This "gw" is to add default route for mvrf which is already there in config_db as part of MGMT_INTERFACE. We just added CLI for supporting that.  This CLI enhancement is only for mvrf and hence other non mvrf options are not being touched.

4) Is the table #1001 reserved for mgmt-vrf, or is it Table# 1 ?
<<Kannan>> Table 5000 is reserved for mvrf. CLI in this PR is already corrected to use 5000 (not 1001).

5) Will be replied in PR472.

6) Is the Curl, Wget, SSH also supported for mgmt-vrf ?
<<Kannan>> Yes, all applications are supported for mgmt-vrf. Those applications do not need any enhancement. Users can do "cgexec -g l3mdev:mgmt COMMAND" for running any application from the device. Similarly, users can establish any application connections to the device from external network as well.

7) Where is the mgmtvrf_ntpq command ?
<<Kannan>> mgmtvrf_ntpq is removed now in this PR463.  Once if this PR463 is merged, same changes will be there in 472.

**Comment by a-barboza** [comment]
@kannankvs: thanks for the replies. For answer 3 above: Will the HLD for management vrf be updated to state the following " As of now, eth0 alone will be part of mvrf; not front panel ports" ? 
Currently the HLD has use cases for In-band management.

**Comment by kannankvs** [comment]
@a-barboza : Yes, we will update it.

File: config/main.py [hunk 1]
```diff
@@ -789,6 +791,86 @@ def del_vlan_member(ctx, vid, interface_name):
     db.set_entry('VLAN_MEMBER', (vlan_name, interface_name), None)
 
 
+
```
Comment (prsunny) [inline config/main.py]: Please remove extra line

File: config/main.py [hunk 2]
```diff
@@ -789,6 +791,86 @@ def del_vlan_member(ctx, vid, interface_name):
...
+	    return None
+	config_db.mod_entry('MGMT_VRF_CONFIG',"vrf_global",{"mgmtVrfEnabled": "true"})
+        cmd="service ntp stop"
+        os.system (cmd)
+        cmd="systemctl restart interfaces-config"
+        os.system (cmd)
+        cmd="service ntp start"
+        os.system (cmd)
+
+
```
Comment (prsunny) [inline config/main.py]: Please remove extra line

File: config/main.py [hunk 3]
```diff
@@ -789,6 +791,86 @@ def del_vlan_member(ctx, vid, interface_name):
...
+#
+# 'vrf' group ('config vrf ...')
+#
+
+@config.group('vrf')
+def vrf():
+    """VRF-related configuration tasks"""
+    pass
+
+
```
Comment (prsunny) [inline config/main.py]: Please remove extra line

File: config/main.py [hunk 4]
```diff
@@ -789,6 +791,86 @@ def del_vlan_member(ctx, vid, interface_name):
...
 
+
+def vrf_add_management_vrf():
+	"""Enable management vrf"""
+
+	config_db = ConfigDBConnector()
+	config_db.connect()
+        entry = config_db.get_entry('MGMT_VRF_CONFIG', "vrf_global")
+        if entry and entry['mgmtVrfEnabled'] == 'true' :
+	    click.echo("ManagementVRF is already Enabled.")
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Done.

File: config/main.py [hunk 5]
```diff
@@ -789,6 +791,86 @@ def del_vlan_member(ctx, vid, interface_name):
...
+	    return None
+	config_db.mod_entry('MGMT_VRF_CONFIG',"vrf_global",{"mgmtVrfEnabled": "true"})
+        cmd="service ntp stop"
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Done. Moved it to a function and provided the required comments.

File: config/main.py [hunk 6]
```diff
@@ -789,6 +791,86 @@ def del_vlan_member(ctx, vid, interface_name):
...
+    """VRF Delete"""
+    if vrfname == 'mgmt' or vrfname == 'management':
+        vrf_delete_management_vrf()
+    else:
+        click.echo("Deletion of data vrf={} is not yet supported".format(vrfname))
+
+
+@config.command('clear_mgmt')
+@click.pass_context
+def clear_mgmt(ctx):
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: clear_mgmt was earlier indented for easy deletion of configured values. Now that we have required "del/remove" functions, it is not required. It is removed now.

File: config/main.py [hunk 7]
```diff
@@ -956,6 +1047,32 @@ def add(ctx, interface_name, ip_addr):
...
+        elif interface_name == 'eth0':
+
+            # Configuring more than 1 IPv4 or more than 1 IPv6 address fails.
+            # Allow only one IPv4 and only one IPv6 address to be configured for IPv6.
+            # If a row already exist, overwrite it (by doing delete and add).
+            mgmtintf_key_list = _get_all_mgmtinterface_keys()
+
+            for key in mgmtintf_key_list:
+                # For loop runs for max 2 rows, once for IPv4 and once for IPv6.
+                if ':' in ip_addr and ':' in key[1]:
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Done. When mvrf was implemented in Feb2019, "ipaddress" module did not have the required path and hence it was unable to import. Now, the "ipaddress" module is available and it is already being used in main.py. It is now used in this function to check the version.

File: config/main.py [hunk 8]
```diff
@@ -991,6 +1108,12 @@ def remove(ctx, interface_name, ip_addr):
         if interface_name.startswith("Ethernet"):
             config_db.set_entry("INTERFACE", (interface_name, ip_addr), None)
             if_table = "INTERFACE"
+        elif interface_name == 'eth0':
+            config_db.set_entry("MGMT_INTERFACE", (interface_name, ip_addr), None)
+            cmd="systemctl restart interfaces-config"
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Done.

File: show/main.py [hunk 9]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
 
     run_command(cmd, display_cmd=verbose)
 
+#
+# 'mgmt-vrf' group ("show mgmt-vrf ...")
+#
+
+@cli.group('mgmt-vrf', invoke_without_command=True)
+@click.pass_context
+def mgmt_vrf(ctx):
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Output of the show commands are added in this PR. All these details will be added into the CLI document once if the PR is merged.

File: show/main.py [hunk 10]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+			# if the mgmtVrfEnabled attribute is configured, check the value
+			# and print Enabled or Disabled accordingly.
+			if 'mgmtVrfEnabled' in mvrf_dict['vrf_global']:
+				if (mvrf_dict['vrf_global']['mgmtVrfEnabled'] == "true"):
+					click.echo("\nManagementVRF : Enabled")
+				else:
+					click.echo("\nManagementVRF : Disabled")
+	
+		click.echo("\nManagement VRF in Linux:")
+		cmd = "sudo ip link show type vrf"
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Yes, it was a mistake. Its corrected now to display only the management VRF related data.

File: show/main.py [hunk 11]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+@mgmt_vrf.command('route')
+def mgmt_vrf_route ():
+	"""Show management VRF routes"""
+	
+	click.echo("\nRoutes in Management VRF Routing Table:")
+	cmd = "sudo ip route show table 5000"
+	run_command(cmd)
+	return None
+
+
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 12]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+@mgmt_vrf.command('addresses')
+def mgmt_vrf_addresses ():
+	"""Show management VRF addresses"""
+	
+	click.echo("\nIP Addresses for interfaces in Management VRF:")
+	cmd = "sudo ip address show mgmt"
+	run_command(cmd)
+	return None
+
+
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 13]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+@cli.group('mgmt-vrf', invoke_without_command=True)
+@click.pass_context
+def mgmt_vrf(ctx):
+
+	"""Show management VRF attributes"""
+
+	if ctx.invoked_subcommand is None:
+		cmd = 'sonic-cfggen -d --var-json "MGMT_VRF_CONFIG"'
+		
+		p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 14]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+					click.echo("\nManagementVRF : Enabled")
+				else:
+					click.echo("\nManagementVRF : Disabled")
+	
+		click.echo("\nManagement VRF in Linux:")
+		cmd = "sudo ip link show type vrf"
+		run_command(cmd)
+
+@mgmt_vrf.command('interfaces')
+def mgmt_vrf_interfaces ():
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Removed this command. Earlier, the idea was to provide separate commands for each of the linux command that were required for debugging. Now the required show is addressed in "show mgmt_vrf" command itself. Hence, removed this command.

File: show/main.py [hunk 15]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+	cmd = "sudo ifconfig eth0"
+	run_command(cmd)
+	return None
+
+@mgmt_vrf.command('route')
+def mgmt_vrf_route ():
+	"""Show management VRF routes"""
+	
+	click.echo("\nRoutes in Management VRF Routing Table:")
+	cmd = "sudo ip route show table 5000"
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 16]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+	run_command(cmd)
+	return None
+
+
+@mgmt_vrf.command('addresses')
+def mgmt_vrf_addresses ():
+	"""Show management VRF addresses"""
+	
+	click.echo("\nIP Addresses for interfaces in Management VRF:")
+	cmd = "sudo ip address show mgmt"
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 17]
```diff
@@ -418,6 +418,93 @@ def ndp(ip6address, iface, verbose):
...
+	"""Show management VRF routes"""
+	
+	click.echo("\nRoutes in Management VRF Routing Table:")
+	cmd = "sudo ip route show table 5000"
+	run_command(cmd)
+	return None
+
+
+@mgmt_vrf.command('addresses')
+def mgmt_vrf_addresses ():
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Removed the confusing command. Like you said "show management_interface address" is good enough to display the configure address for eth0.

File: show/main.py [hunk 18]
```diff
@@ -1443,8 +1530,24 @@ def bgp(verbose):
 @click.option('--verbose', is_flag=True, help="Enable verbose output")
 def ntp(verbose):
     """Show NTP information"""
-    cmd = "ntpq -p -n"
-    run_command(cmd, display_cmd=verbose)
+    ntpcmd = "ntpq -p -n"
+    if ctx.invoked_subcommand is None:
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: config/main.py (hunk 1 above)
Comment (kannankvs) [reply config/main.py]: Done. Removed extra lines between the new definitions that got added for mvrf.

File: config/main.py (hunk 2 above)
Comment (kannankvs) [reply config/main.py]: Done

File: config/main.py (hunk 3 above)
Comment (kannankvs) [reply config/main.py]: Done

**Comment by kannankvs** [comment]
@prsunny : Comments are addressed. Request you to review and merge.

File: show/main.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: While resolving conflict, change got overwritten. Its resolved now. Apologies.

File: config/main.py [hunk 19]
```diff
@@ -889,6 +891,79 @@ def del_vlan_member(ctx, vid, interface_name):
...
+    """Enable management vrf"""
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    entry = config_db.get_entry('MGMT_VRF_CONFIG', "vrf_global")
+    if entry and entry['mgmtVrfEnabled'] == 'true' :
+        click.echo("ManagementVRF is already Enabled.")
+        return None
+    config_db.mod_entry('MGMT_VRF_CONFIG',"vrf_global",{"mgmtVrfEnabled": "true"})
+    # To move eth0 to management VRF, interfaces-config service has be to be
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Actually the comments are slightly different for enable and disable conditions. That is why it was provided at this place. Now, the comments are moved inside the function and reworded to suit for both enable and disable conditions. By multi-line, we assume you meant the usage of """, which is also done now.

File: config/main.py [hunk 20]
```diff
@@ -1112,6 +1187,21 @@ def speed(ctx, interface_name, interface_speed, verbose):
...
 
+def _get_all_mgmtinterface_keys():
+    """Returns list of strings containing mgmt interface keys 
+    """
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    return config_db.get_table('MGMT_INTERFACE').keys()
+
+
+def restart_interfaces_config_ntp_config():
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: There is a difference. One is restarting the "ntp-config" service and the other is restarting "ntp" service.  "ntp-config" service will regenerate the ntp.conf file, which is required when the IP address for eth0 is modified. Same is not required when VRF is enabled or disabled.

File: config/main.py [hunk 21]
```diff
@@ -1112,6 +1187,21 @@ def speed(ctx, interface_name, interface_speed, verbose):
         command += " -vv"
     run_command(command, display_cmd=verbose)
 
+def _get_all_mgmtinterface_keys():
+    """Returns list of strings containing mgmt interface keys 
+    """
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    return config_db.get_table('MGMT_INTERFACE').keys()
+
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: While resolving conflict, it got missed out. Its corrected now.

File: config/main.py [hunk 22]
```diff
@@ -1147,6 +1238,38 @@ def add(ctx, interface_name, ip_addr):
...
+            # Allow only one IPv4 and only one IPv6 address to be configured for IPv6.
+            # If a row already exist, overwrite it (by doing delete and add).
+            mgmtintf_key_list = _get_all_mgmtinterface_keys()
+
+            for key in mgmtintf_key_list:
+                # For loop runs for max 2 rows, once for IPv4 and once for IPv6.
+                # No need to capture the exception since the ip_addr is already validated earlier
+                ip_input = ipaddress.ip_interface(ip_addr)
+                current_ip = ip = ipaddress.ip_interface(key[1])
+                if (ip_input.version == 6) and (current_ip.version == 6):
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: yes, it will cover. Changed.

File: config/main.py [hunk 23]
```diff
@@ -1147,6 +1238,38 @@ def add(ctx, interface_name, ip_addr):
...
+                elif (ip_input.version != 6) and (current_ip.version != 6):
+                    # If user has configured IPv4 address and the already available row is also IPv4, delete it here.
+                    config_db.set_entry("MGMT_INTERFACE", ("eth0", key[1]), None)
+
+            # Set the new row with new value
+            if not gw:
+                config_db.set_entry("MGMT_INTERFACE", (interface_name, ip_addr), {"NULL": "NULL"})
+            else:
+                config_db.set_entry("MGMT_INTERFACE", (interface_name, ip_addr), {"gwaddr": gw})
+            # Restart the interfaces-config service which regenerates the
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Done.

File: config/main.py [hunk 24]
```diff
@@ -1190,6 +1313,15 @@ def remove(ctx, interface_name, ip_addr):
             else:
                 config_db.set_entry("INTERFACE", (interface_name, ip_addr), None)
                 if_table = "INTERFACE"
+        elif interface_name == 'eth0':
+            config_db.set_entry("MGMT_INTERFACE", (interface_name, ip_addr), None)
+            # Restart the interfaces-config service which regenerates the
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply config/main.py]: Done.

File: show/main.py [hunk 25]
```diff
@@ -433,6 +433,76 @@ def ndp(ip6address, iface, verbose):
...
     run_command(cmd, display_cmd=verbose)
 
+def is_mgmt_vrf_enabled(ctx):
+    """Check if management VRF is enabled"""
+    if ctx.invoked_subcommand is None:
+        cmd = 'sonic-cfggen -d --var-json "MGMT_VRF_CONFIG"'
+
+        p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        res = p.communicate()
+        if p.returncode == 0 :
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 26]
```diff
@@ -433,6 +433,76 @@ def ndp(ip6address, iface, verbose):
...
+            p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            mvrf_dict = json.loads(p.stdout.read())
+
+            # if the mgmtVrfEnabled attribute is configured, check the value
+            # and print Enabled or Disabled accordingly.
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 27]
```diff
@@ -433,6 +433,76 @@ def ndp(ip6address, iface, verbose):
...
+#
+
+@cli.group('mgmt-vrf', invoke_without_command=True)
+@click.argument('routes', required=False)
+@click.pass_context
+def mgmt_vrf(ctx,routes):
+    """Show management VRF attributes"""
+
+    if is_mgmt_vrf_enabled(ctx) is False:
+        click.echo("\nKVSK:ManagementVRF : Disabled")
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

File: show/main.py [hunk 28]
```diff
@@ -433,6 +433,76 @@ def ndp(ip6address, iface, verbose):
...
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    header = ['IFNAME', 'IP Address', 'PrefixLen',]
+    body = []
+
+    # Fetching data from config_db for MGMT_INTERFACE
+    mgmt_ip_data = config_db.get_table('MGMT_INTERFACE')
+    for key in natsorted(mgmt_ip_data.keys()):
+        click.echo("Management IP address = {0}".format(key[1]))
+        click.echo("Management NetWork Default Gateway = {0}".format(mgmt_ip_data[key]['gwaddr']))
```
[the remark this answers is not in the document]
Comment (kannankvs) [reply show/main.py]: Done.

**Comment by kannankvs** [comment]
@prsunny :  Addressed the new comments as well.

File: config/main.py [hunk 29]
```diff
@@ -6,6 +6,8 @@
 import json
 import subprocess
 import netaddr
+import logging
+import logging.handlers
```
Comment (jleveque) [inline config/main.py]: These imports don't appear to be used. Remove?

File: config/main.py [hunk 30]
```diff
@@ -889,6 +891,76 @@ def del_vlan_member(ctx, vid, interface_name):
...
+@config.group('vrf')
+def vrf():
+    """VRF-related configuration tasks"""
+    pass
+
+@vrf.command('add')
+@click.argument('vrfname', metavar='<vrfname>. Type mgmt for management VRF', required=True)
+@click.pass_context
+def vrf_add (ctx, vrfname):
+    """VRF ADD"""
```
Comment (jleveque) [inline config/main.py]: Docstring is used as help message. Can you make it into a brief, clear statement?

File: config/main.py [hunk 31]
```diff
@@ -889,6 +891,76 @@ def del_vlan_member(ctx, vid, interface_name):
...
+    if vrfname == 'mgmt' or vrfname == 'management':
+        vrf_add_management_vrf()
+    else:
+        click.echo("Creation of data vrf={} is not yet supported".format(vrfname))
+
+@vrf.command('del')
+@click.argument('vrfname', metavar='<vrfname>. Type mgmt for management VRF', required=False)
+@click.pass_context
+def vrf_del (ctx, vrfname):
+    """VRF Delete"""
```
Comment (jleveque) [inline config/main.py]: Docstring is used as help message. Can you make it into a brief, clear statement?

Comment (kannankvs) [reply config/main.py]: Done.

File: config/main.py (hunk 30 above)
Comment (kannankvs) [reply config/main.py]: Done.

File: config/main.py (hunk 29 above)
Comment (kannankvs) [reply config/main.py]: @jleveque : Done. All changes are completed.

**Comment by wendani** [comment]
Shall "MGMT_INTERFACE|eth0" be updated into the vrf "mgmt" in the CONFIG_DB when doing `sudo config vrf add mgmt` according to the VRF schema convention?

```
    "MGMT_INTERFACE": {
        "eth0": {
            "vrf_name": "mgmt"
        }
        "eth0|10.2.17.103/23": {
            "gwaddr": "10.2.16.1"
        }
    },
```

**Comment by prsunny** [comment]
Thats right Wenda

**Pull request merged**